### PR TITLE
Use `pynvjitlink` for CUDA 12+ MVC

### PR DESCRIPTION
**Pull request by brandon-b-miller**
Fixes https://github.com/rapidsai/cudf/issues/12822

This PR provides minor version compatibility in the CUDA 12.x range through `nvjitlink` via the preliminary [nvjiitlink python binding](https://github.com/gmarkall/nvjitlink). Thus far this PR merely leverages a local installation of the library and should not be merged until `nvjitlink` is hosted on `conda-forge` and cuDF's dependencies are adjusted accordingly, likely as part of this PR.

### CODE REVIEW

**Comment by jakirkham** [comment]
Added a "Fixes" note in the OP. @brandon-b-miller please double check if that is correct

File: python/cudf/cudf/utils/_numba.py
Comment (bdice) [inline python/cudf/cudf/utils/_numba.py:14]: I think you can use the more generic `ImportError` (parent class of `ModuleNotFoundError`).
```suggestion
except ImportError:
```

File: python/cudf/cudf/utils/_numba.py
Comment (bdice) [inline python/cudf/cudf/utils/_numba.py:177]: Can we add 12.3 here, too?

File: python/cudf/cudf/tests/test_numba_import.py
Comment (bdice) [inline python/cudf/cudf/tests/test_numba_import.py:25]: This seems to be unconditionally using CUDA 11 patching logic in tests. We should use CUDA 12 patching logic when on CUDA 12, right?

**Review by bdice — CHANGES_REQUESTED** [review]

File: python/cudf/cudf/tests/test_numba_import.py
Comment (brandon-b-miller) [reply python/cudf/cudf/tests/test_numba_import.py:25]: I think this whole test skips in a CUDA 12 environment based on the below
```
@pytest.mark.skipif(
    not IS_CUDA_11, reason="Minor Version Compatibility test for CUDA 11"
)
```

Comment (bdice) [reply python/cudf/cudf/tests/test_numba_import.py:25]: The MVC tests should be enabled, even if we don't have an MVC-triggering environment in CI with the drivers/runtimes we use. We want to be able to run `pytest` on a system requiring MVC and know that it's being covered by cudf.

Comment (bdice) [reply python/cudf/cudf/tests/test_numba_import.py:25]: It's okay to have them be skipped if on CUDA 12 `and` MVC is needed `and` pynvjitlink isn't available.

Comment (brandon-b-miller) [reply python/cudf/cudf/tests/test_numba_import.py:25]: Ok, that makes sense. I'm thinking I'll rename this file `test_mvc` and have a separate test for cuda 11 and cuda 12. updates to follow.

Comment (brandon-b-miller) [reply python/cudf/cudf/tests/test_numba_import.py]: did some refactoring in https://github.com/rapidsai/cudf/pull/13650/commits/35b7e59e002c7c1bf07eb61063cb86a7c61c6577

File: python/cudf/cudf/tests/test_mvc.py
Comment (bdice) [inline python/cudf/cudf/tests/test_mvc.py:19]:
```suggestion
    HAVE_PYNVJITLINK = True
```

File: python/cudf/cudf/utils/_numba.py
Comment (bdice) [inline python/cudf/cudf/utils/_numba.py:16]: Do we expect that future major versions of CUDA will also use pynvjitlink? Do we want the naming of these functions to reflect "nvjitlink" rather than "cuda_12"?
```suggestion
    def patch_numba_linker_pynvjitlink():
```

File: python/cudf/cudf/utils/_numba.py
Comment (bdice) [inline python/cudf/cudf/utils/_numba.py:81]: Maybe this should be named by the package being used rather than the major version? But I'm okay with this one saying CUDA 11 since it really is specific to that major version.
```suggestion
def patch_numba_linker_cubinlinker():
```

**Review by bdice — COMMENTED** [review]
A few cleanups -- but this is looking close to right. Have you verified that the tests run locally as expected with pynvjitlink?

**Comment by brandon-b-miller** [comment]
> A few cleanups -- but this is looking close to right. Have you verified that the tests run locally as expected with pynvjitlink?

For me at least tests pass locally building from source with a 12.0 driver and 12.3 CTK. I'll also do a round of follow up testing on the tip of 23.12 after the merge.

File: python/cudf/cudf/utils/_numba.py
Comment (bdice) [inline python/cudf/cudf/utils/_numba.py:19]: Depending on the timing of pynvjitlink being shipped, we'll want to do one of the following:
- If pynvjitlink isn't shipped by the time we release, we should remove this note and say something like before (explain the MVC problem and direct users to match their driver and toolkit)
- If pynvjitlink _is_ available in time, we should give the full install command, like `pip install pynvjitlink --extra-index-url=https://pypi.nvidia.com`

**Review by bdice — APPROVED** [review]
Approving with one request for a final change based on the status/timeline of pynvjitlink wheels being published: https://github.com/rapidsai/cudf/pull/13650#discussion_r1398487365

File: python/cudf/cudf/utils/_numba.py
Comment (brandon-b-miller) [reply python/cudf/cudf/utils/_numba.py:19]: How do you feel about keeping the original warning to be safe then. The changes will still allow `pynvjitlink` to work once it's out, but we won't document it or direct users to it in any official way until then.

Comment (bdice) [reply python/cudf/cudf/utils/_numba.py:19]: Yes, do that. We can change the warning during code freeze if we ship pynvjitlink in time.

**Review by shwina — APPROVED** [review]

**Pull request merged**
